### PR TITLE
feat: litecoin

### PR DESCRIPTION
**Pull request by kaladinlight**
- also add/improves utxo abstractions and shared types/logic

### CODE REVIEW

File: packages/caip/src/adapters/coingecko/generated/bip122_000000000019d6689c085ae165831e93/adapter.json [hunk 1]
```diff
@@ -1 +1 @@
-{"bip122:000000000019d6689c085ae165831e93/slip44:0":"bitcoin"}
\ No newline at end of file
+{}
```
Comment (0xdef1cafe) [inline packages/caip/src/adapters/coingecko/generated/bip122_000000000019d6689c085ae165831e93/adapter.json:1]: where did he go?

File: packages/caip/src/utils.ts [hunk 2]
```diff
@@ -51,55 +34,13 @@ export const getFeeAssetIdFromAssetId = (assetId: AssetId): AssetId | undefined
...
-  const chainReference = CHAIN_REFERENCE.OsmosisMainnet
-  const assetId = toAssetId({
-    chainNamespace,
-    chainReference,
-    assetNamespace: 'slip44',
-    assetReference: ASSET_REFERENCE.Osmosis
-  })
-  return { [assetId]: 'osmosis' }
-}
+export const bitcoinAsset = { [constants.btcAssetId]: 'bitcoin' }
```
Comment (0xdef1cafe) [inline packages/caip/src/utils.ts:39]: `bitcoinAssetMap` or `bitcoinAssetData` - flip a coin

File: packages/chain-adapters/src/cosmossdk/cosmos/CosmosChainAdapter.ts [hunk 3]
```diff
@@ -173,7 +169,7 @@ export class ChainAdapter extends CosmosSdkBaseAdapter<KnownChainIds.CosmosMainn
       const txToSign: CosmosSignTx = {
         addressNList,
         tx: utx,
-        chain_id: CHAIN_REFERENCE.CosmosHubMainnet,
+        chain_id: this.chainId,
```
Comment (0xdef1cafe) [inline packages/chain-adapters/src/cosmossdk/cosmos/CosmosChainAdapter.ts:172]: regression

File: packages/chain-adapters/src/cosmossdk/cosmos/CosmosChainAdapter.ts [hunk 4]
```diff
@@ -245,7 +241,7 @@ export class ChainAdapter extends CosmosSdkBaseAdapter<KnownChainIds.CosmosMainn
...
```
Comment (0xdef1cafe) [inline packages/chain-adapters/src/cosmossdk/cosmos/CosmosChainAdapter.ts:244]: regression

File: packages/chain-adapters/src/cosmossdk/cosmos/CosmosChainAdapter.ts [hunk 5]
```diff
@@ -316,7 +312,7 @@ export class ChainAdapter extends CosmosSdkBaseAdapter<KnownChainIds.CosmosMainn
...
```
Comment (0xdef1cafe) [inline packages/chain-adapters/src/cosmossdk/cosmos/CosmosChainAdapter.ts:315]: regression

File: packages/chain-adapters/src/cosmossdk/cosmos/CosmosChainAdapter.ts [hunk 6]
```diff
@@ -381,7 +377,7 @@ export class ChainAdapter extends CosmosSdkBaseAdapter<KnownChainIds.CosmosMainn
...
```
Comment (0xdef1cafe) [inline packages/chain-adapters/src/cosmossdk/cosmos/CosmosChainAdapter.ts:380]: regression

File: packages/chain-adapters/src/cosmossdk/cosmos/CosmosChainAdapter.ts [hunk 7]
```diff
@@ -460,7 +456,7 @@ export class ChainAdapter extends CosmosSdkBaseAdapter<KnownChainIds.CosmosMainn
...
```
Comment (0xdef1cafe) [inline packages/chain-adapters/src/cosmossdk/cosmos/CosmosChainAdapter.ts:459]: regression

File: packages/unchained-client/generator/openapitools-dev.json [hunk 8]
```diff
@@ -5,18 +5,18 @@
...
+        "output": "#{cwd}/src/generated/avalanche",
         "reservedWordsMappings": { "in": "in" },
         "additionalProperties": {
           "supportsES6": "true",
           "useSingleRequestParameter": true
         }
       },
       "bitcoin": {
-        "inputSpec": "https://dev-api.bitcoin.shapeshift.com/swagger.json",
+        "inputSpec": "http://api.bitcoin.localhost/swagger.json",
```
Comment (0xdef1cafe) [inline packages/unchained-client/generator/openapitools-dev.json:19]: fix

File: packages/unchained-client/src/utxo/bitcoin/parser/index.ts [hunk 9]
```diff
@@ -0,0 +1,18 @@
...
+import { Tx } from '../../../generated/bitcoin'
+import { BaseTransactionParser, BaseTransactionParserArgs } from '../../parser'
+
+export type TransactionParserArgs = BaseTransactionParserArgs
+
+export class TransactionParser extends BaseTransactionParser<Tx> {
+  constructor(args: TransactionParserArgs) {
+    super(args)
+
+    this.assetId = toAssetId({
```
Comment (0xdef1cafe) [inline packages/unchained-client/src/utxo/bitcoin/parser/index.ts:12]: btcAssetId

File: packages/unchained-client/src/utxo/dogecoin/parser/index.ts [hunk 10]
```diff
@@ -0,0 +1,18 @@
...
+import { Tx } from '../../../generated/dogecoin'
+import { BaseTransactionParser, BaseTransactionParserArgs } from '../../parser'
+
+export type TransactionParserArgs = BaseTransactionParserArgs
+
+export class TransactionParser extends BaseTransactionParser<Tx> {
+  constructor(args: TransactionParserArgs) {
+    super(args)
+
+    this.assetId = toAssetId({
```
Comment (0xdef1cafe) [inline packages/unchained-client/src/utxo/dogecoin/parser/index.ts:12]: dogeAssetId

File: packages/unchained-client/src/utxo/index.ts [hunk 11]
```diff
@@ -0,0 +1,5 @@
+export type { utxo as types } from '@shapeshiftoss/common-api'
+export { BaseTransactionParser } from './parser'
+
+export * as bitcoin from './bitcoin'
+export * as dogecoin from './dogecoin'
```
Comment (0xdef1cafe) [inline packages/unchained-client/src/utxo/index.ts:5]: litecoin?

File: packages/unchained-client/src/utxo/litecoin/parser/index.ts [hunk 12]
```diff
@@ -0,0 +1,18 @@
...
+import { Tx } from '../../../generated/litecoin'
+import { BaseTransactionParser, BaseTransactionParserArgs } from '../../parser'
+
+export type TransactionParserArgs = BaseTransactionParserArgs
+
+export class TransactionParser extends BaseTransactionParser<Tx> {
+  constructor(args: TransactionParserArgs) {
+    super(args)
+
+    this.assetId = toAssetId({
```
Comment (0xdef1cafe) [inline packages/unchained-client/src/utxo/litecoin/parser/index.ts:12]: ltcAssetId

File: packages/chain-adapters/src/utxo/UtxoBaseAdapter.ts [hunk 13]
```diff
@@ -42,67 +41,85 @@ import { bnOrZero } from '../utils/bignumber'
...
+  defaultBIP44Params: BIP44Params
+  defaultUtxoAccountType: UtxoAccountType
+  supportedChainIds: Array<ChainId>
+  supportedAccountTypes: Array<UtxoAccountType>
+  chainId: UtxoChainId
+}
+
+export abstract class UtxoBaseAdapter<T extends UtxoChainId> implements IChainAdapter<T> {
+  protected readonly chainId: UtxoChainId
+  protected readonly coinName: string
```
Comment (0xdef1cafe) [inline packages/chain-adapters/src/utxo/UtxoBaseAdapter.ts:71]: don't address - wtf is this even for any mroe?

File: packages/caip/src/utils.ts (hunk 2 above)
Comment (kaladinlight) [reply packages/caip/src/utils.ts]: addressed in https://github.com/shapeshift/lib/pull/914/commits/080c21f5f8f5aa8c96b55e31126c9b828da30d52

File: packages/chain-adapters/src/cosmossdk/cosmos/CosmosChainAdapter.ts (hunk 3 above)
Comment (kaladinlight) [reply packages/chain-adapters/src/cosmossdk/cosmos/CosmosChainAdapter.ts]: addressed in https://github.com/shapeshift/lib/pull/914/commits/080c21f5f8f5aa8c96b55e31126c9b828da30d52

File: packages/chain-adapters/src/cosmossdk/cosmos/CosmosChainAdapter.ts (hunk 4 above)
Comment (kaladinlight) [reply packages/chain-adapters/src/cosmossdk/cosmos/CosmosChainAdapter.ts]: addressed in https://github.com/shapeshift/lib/pull/914/commits/080c21f5f8f5aa8c96b55e31126c9b828da30d52

File: packages/chain-adapters/src/cosmossdk/cosmos/CosmosChainAdapter.ts (hunk 5 above)
Comment (kaladinlight) [reply packages/chain-adapters/src/cosmossdk/cosmos/CosmosChainAdapter.ts]: addressed in https://github.com/shapeshift/lib/pull/914/commits/080c21f5f8f5aa8c96b55e31126c9b828da30d52

File: packages/chain-adapters/src/cosmossdk/cosmos/CosmosChainAdapter.ts (hunk 6 above)
Comment (kaladinlight) [reply packages/chain-adapters/src/cosmossdk/cosmos/CosmosChainAdapter.ts]: addressed in https://github.com/shapeshift/lib/pull/914/commits/080c21f5f8f5aa8c96b55e31126c9b828da30d52

File: packages/chain-adapters/src/cosmossdk/cosmos/CosmosChainAdapter.ts (hunk 7 above)
Comment (kaladinlight) [reply packages/chain-adapters/src/cosmossdk/cosmos/CosmosChainAdapter.ts]: addressed in https://github.com/shapeshift/lib/pull/914/commits/080c21f5f8f5aa8c96b55e31126c9b828da30d52

File: packages/unchained-client/generator/openapitools-dev.json (hunk 8 above)
Comment (kaladinlight) [reply packages/unchained-client/generator/openapitools-dev.json]: addressed in https://github.com/shapeshift/lib/pull/914/commits/080c21f5f8f5aa8c96b55e31126c9b828da30d52

File: packages/unchained-client/src/utxo/bitcoin/parser/index.ts (hunk 9 above)
Comment (kaladinlight) [reply packages/unchained-client/src/utxo/bitcoin/parser/index.ts]: addressed in https://github.com/shapeshift/lib/pull/914/commits/080c21f5f8f5aa8c96b55e31126c9b828da30d52

File: packages/unchained-client/src/utxo/dogecoin/parser/index.ts (hunk 10 above)
Comment (kaladinlight) [reply packages/unchained-client/src/utxo/dogecoin/parser/index.ts]: addressed in https://github.com/shapeshift/lib/pull/914/commits/080c21f5f8f5aa8c96b55e31126c9b828da30d52

File: packages/unchained-client/src/utxo/index.ts (hunk 11 above)
Comment (kaladinlight) [reply packages/unchained-client/src/utxo/index.ts:5]: addressed in https://github.com/shapeshift/lib/pull/914/commits/080c21f5f8f5aa8c96b55e31126c9b828da30d52

File: packages/unchained-client/src/utxo/litecoin/parser/index.ts (hunk 12 above)
Comment (kaladinlight) [reply packages/unchained-client/src/utxo/litecoin/parser/index.ts]: addressed in https://github.com/shapeshift/lib/pull/914/commits/080c21f5f8f5aa8c96b55e31126c9b828da30d52

File: packages/chain-adapters/src/utxo/UtxoBaseAdapter.ts (hunk 13 above)
Comment (kaladinlight) [reply packages/chain-adapters/src/utxo/UtxoBaseAdapter.ts:71]: hdwallet goodness

**Review by 0xdef1cafe — APPROVED** [review]
not as scary as it looks - mostly moving things around. 

merging as we currently can't build `main` and may need a follow up PR here to get package versions in sync and linking locally

**Pull request merged**